compiler/rustc_middle/src/query/mod.rs

```diff
@@ -118,6 +118,16 @@ rustc_queries! {
         desc { "perform lints prior to macro expansion" }
     }
 
+    /// Useful for the implementation of `std::env!`, `proc-macro`s change
+    /// detection and target-dependent compiler flags.
+    ///
+    /// Will emit an error and return `None` if the variable is not UTF-8.
+    query env_var(key: Symbol) -> Option<Symbol> {
```

Comment:
Is it possible to use `OsStr(ing)`s in query inputs and outputs? What are the obstacles?
Many places in the compiler use `env::var_os`.

Also, `Symbol`s shouldn't be needed here, you can take a `OsStr`/`str` and return `&'tcx OsString`/`&'tcx String` like `env::var(_os)` returns.

Reply:
`OsStr` isn't really set up with the query system, so we'd have to do that first. I've done parts of it in the last commit, but it also needs to be `Key`, which I'm unsure would be correct? `&str` for example isn't `Key` either.

Also, returning `&'tcx OsStr` requires that the environment variables are actually stored inside `TyCtxt<'tcx>`, which I'm unsure how to do wrt. lifetimes?

Reply:
>Also, returning `&'tcx OsStr` requires that the environment variables are actually stored inside `TyCtxt<'tcx>`, which I'm unsure how to do wrt. lifetimes?

Things can be stored in inside `TyCtxt` using the `tcx.arena.alloc*` methods.
A lot of queries return references to something allocated on tcx arenas.

Reply:
>I've done parts of it in the last commit, but it also needs to be `Key`, which I'm unsure would be correct?

I don't know what exactly needs to be implemented to support OS strings in queries.
I'll need to try doing that myself to answer questions like this, for which I probably don't have time at the moment.

Reply:
I got it to work, thanks for the tips!